service/history/replicationTaskExecutor.go

```diff
@@ -0,0 +1,341 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination replicationTaskExecutor_mock.go -self_package github.com/uber/cadence/service/history
+
+package history
+
+import (
+	"context"
+
+	"github.com/uber/cadence/.gen/go/history"
+	r "github.com/uber/cadence/.gen/go/replicator"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/xdc"
+)
+
+type (
+	replicationTaskExecutor interface {
+		execute(sourceCluster string, replicationTask *r.ReplicationTask, forceApply bool) (int, error)
+	}
+
+	replicationTaskExecutorImpl struct {
+		currentCluster      string
+		domainCache         cache.DomainCache
+		nDCHistoryResender  xdc.NDCHistoryResender
+		historyRereplicator xdc.HistoryRereplicator
+		historyEngine       Engine
+
+		metricsClient metrics.Client
+		logger        log.Logger
+	}
+)
+
+// newReplicationTaskExecutor creates an replication task executor
+// The executor uses by 1) DLQ replication task handler 2) history replication task processor
+func newReplicationTaskExecutor(
+	currentCluster string,
+	domainCache cache.DomainCache,
+	nDCHistoryResender xdc.NDCHistoryResender,
+	historyRereplicator xdc.HistoryRereplicator,
+	historyEngine Engine,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) replicationTaskExecutor {
+	return &replicationTaskExecutorImpl{
+		currentCluster:      currentCluster,
+		domainCache:         domainCache,
+		nDCHistoryResender:  nDCHistoryResender,
+		historyRereplicator: historyRereplicator,
+		historyEngine:       historyEngine,
+		metricsClient:       metricsClient,
+		logger:              logger,
+	}
+}
+
+func (e *replicationTaskExecutorImpl) execute(
+	sourceCluster string,
+	replicationTask *r.ReplicationTask,
+	forceApply bool,
+) (int, error) {
+
+	var err error
+	var scope int
+	switch replicationTask.GetTaskType() {
+	case r.ReplicationTaskTypeSyncShardStatus:
+		// Shard status will be sent as part of the Replication message without kafka
+	case r.ReplicationTaskTypeSyncActivity:
+		scope = metrics.SyncActivityTaskScope
+		err = e.handleActivityTask(replicationTask, forceApply)
+	case r.ReplicationTaskTypeHistory:
+		scope = metrics.HistoryReplicationTaskScope
+		err = e.handleHistoryReplicationTask(sourceCluster, replicationTask, forceApply)
+	case r.ReplicationTaskTypeHistoryMetadata:
+		// Without kafka we should not have size limits so we don't necessary need this in the new replication scheme.
```

Comment:
Use some default scope here? Otherwise metrics like processing latency, attempt, etc. will be emitted to some unknown scope